psydac/linalg/block.py

```diff
@@ -292,6 +292,14 @@ def __setitem__(self, key, value):
         assert isinstance(value, Vector)
         self._blocks[key] = value
 
+    #...
+    def axpy(self, v2, a):
+        assert isinstance(v2, BlockVector)
+        assert v2._space is self._space
+        for i in range(len(self.blocks)):
+            self.blocks[i].axpy(v2.blocks[i], a)
+        self._sync = self._sync and v2._sync
```

Comment:
`axpy` is a standard BLAS function which has signature`axpy(a, x, y)` (excluding those C arguments that are useless in Python).

Therefore, if we want to use the same name for this function we should also try to keep the same signature. For this reason it would be natural to make `axpy` a method of the `VectorSpace` base class, to be subclassed by `StencilVectorSpace` and `BlockVectorSpace`:

```python
class VectorSpace:
    ...
    @abstractmethod
    def axpy(self, a, x, y):
        """ Docstring. """
```

If we want to add a method to the `Vector` base class instead, then we should change its name to something like `mul_iadd`:

```python
class Vector:
    ...
    @abstractmethod
    def mul_iadd(self, a, x):
        """ Docstring. """
```

Reply:
This PR create the two function
[9080230](https://github.com/pyccel/psydac/pull/310/commits/9080230dd760c0c6884c163acd3e2e1d1d3aa46c)

Reply:
I guess I prefer mul_iadd.

Using axpy, we have to include the VectorSpace. But in certain situations, we create vectors without explicitly using a VectorSpace, for example when projecting a continuous function into a discrete space:
```
domain  = logical_domain
derham  = Derham(domain)

domain_h = discretize(domain, ncells=ncells, periodic=periodic, comm=MPI.COMM_WORLD)
derham_h = discretize(derham, domain_h, degree=degree)

P0, P1, P2, P3  = derham_h.projectors(nquads=[5,5,5])

a_1 = lambda x, y, z : ...
a_2 = lambda x, y, z : ...
a_3 = lambda x, y, z : ...
a   = (a_1, a_2, a_3)

a_coeff = P1(a).coeffs 
``` 
If I wanted to add a multiple of a different coefficient vector to `a_coeff`, I'd first have to explicitly mention `V1.vector_space`, which was not necessary previously.

Reply:
Tom has provided both functions `VectorSpace.axpy()` and `Vector.mul_iadd()`.